RavenDB-12025 High CPU usage in idle state

karmeli87 opened the pull request:
This was usually the case when we had many indexes and/or databases (several hundreds in total).

- Reduce network traffic by sending an empty report to the supervisor if no changes were made to any of the database's envs.
- Reduce openning of read transactions and disk access when no change was made to a specific index or database.
- Reduce DatabaseRecord deserialization costs when the database is disabled by accessing the blittable directly.
- Bump the maintenance TCP protocol version and rename `Heartbeats` to `Maintenance`.

karmeli87 committed:
RavenDB-12025 High CPU usage in idle state

This was usually the case when we had many indexes and/or databases (several hundreds in total).

- Reduce network traffic by sending an empty report to the supervisor if no changes were made to any of the database's envs.
- Reduce openning of read transactions and disk access when no change was made to a specific index or database.
- Reduce DatabaseRecord deserialization costs when the database is disabled by accessing the blittable directly.
- Bump the maintenance TCP protocol version and rename `Heartbeats` to `Maintenance`.

karmeli87 commented:
If we want to port this change to 4.0 version also, then it would be preferable to change the TCP version to something like 40_700.

ayende commented on src/Raven.Server/ServerWide/Maintenance/ClusterMaintenanceSupervisor.cs:
```diff
@@ -248,7 +249,16 @@ private void ListenToMaintenanceWorker()
+                                    var dbReport = report.Report[name];
+                                    if (dbReport.Status == DatabaseStatus.NoChange)
+                                    {
+                                        report.Report[name] = _lastSuccessfulReceivedReport.Report[name];
```
Is this correct? It looks like we'll miss out on the things that are updated, no? Like the replication stats?

karmeli87 replied:
Fixed.

ayende commented on src/Raven.Server/ServerWide/Maintenance/ClusterMaintenanceSupervisor.cs:
```diff
@@ -248,7 +249,16 @@ private void ListenToMaintenanceWorker()
                                 var report = BuildReport(rawReport);
                                 timeoutEvent.Defer(_parent._leaderClusterTag);
-
+                                foreach (var name in report.Report.Keys.ToList())
```
Better to avoid this allocation (should be false most of the time when we are active) and only allocate if we indeed get a no chance notification.
For that matter, since we are single threaded, we can just have a separate list of the modified values that we'll then set after the `foreach`

karmeli87 replied:
Fixed.

karmeli87 committed:
RavenDB-12025 fix PR comment

ayende committed:
Minor fixes

* Use a local variable to avoid multiple dictionary checks
* Check if the old value exists (it might not) if the node reports that it is unchanged
* We need to clear the list of unchanged reports as soon as we are done using it.

ayende commented on src/Raven.Server/ServerWide/Maintenance/ClusterMaintenanceSupervisor.cs:
```diff
@@ -269,6 +275,35 @@ private void ListenToMaintenanceWorker()
+                foreach (var dbReport in unchangedReports)
+                {
+                    var dbName = dbReport.Name;
+                    if(_lastSuccessfulReceivedReport.TryGetValue(name, out var previous) == false)
```
@karmeli87 Need to see if the caller code can handle the case where we get the unchanged directly.

karmeli87 committed:
RavenDB-12025 fix compilation

karmeli87 committed:
RavenDB-12025 High CPU usage in idle state

- on `NoChange` status, update the report with the recent successful time
- on we should not send `NoChange` status.
- avoid unnecessary write-tx on replication failure

karmeli87 committed:
RavenDB-12025 Revert the name to `Heartbeats` since it breaks backward comp.

ayende commented:
Please check the failing tests

karmeli87 commented:
I think they fail regardless to this PR
http://issues.hibernatingrhinos.com/issue/RavenDB-11917
http://issues.hibernatingrhinos.com/issue/RavenDB-11968
